### PR TITLE
Patch 4

### PR DESCRIPTION
**Pull request by clarelsy**
Fixed bugs and polish.

### CODE REVIEW

File: ch13/ex13_34_36_37.cpp [hunk 1]
```diff
@@ -48,7 +44,15 @@ void Message::remove_from_Folders()
...
-    folders.clear();
+    // The book added one line here: folders.clear(); but I think it is redundant and more importantly, it will cause a bug:
+    // - In Message::operator=, in the case of self-assignment, it first calls remove_from_Folders() and its folders.clear() 
+    //   clears the data member of lhs(rhs), and there is no way we can assign it back to lhs.
+    //   Refer to: http://stackoverflow.com/questions/29308115/protection-again-self-assignment
+    // - Why is it redundant? As its analogous function Message::add_to_Folders(), Message::remove_from_Folders() should ONLY
+    //   take care of the bookkeeping in Folders but not touch the Message's own data members - makes it much clearer and easier
+    //   to use. As you can see in the 2 places where we call Message::remove_from_Folders(): in Message::operator=, folders.clear()
+    //   introduces a bug as illustrated above; in the destructor ~Message(), the member "folders" will be destroyed anyways, why do
+    //   we need to clear it first?
```
Comment (pezy) [inline ch13/ex13_34_36_37.cpp]: Sorry for my poor understanding, could you please show the code for reappear the bug your mentioned?

**Comment by clarelsy** [comment]
As a simple example, with your original code (same as the book), you called folders.clear() in Message::remove_from_Folders(). Test it with the following main():

int main()
{
	Message m1("Message1");
	Folder f1;

	m1.save(f1);
	m1.print_debug(); // Output: Message1
	f1.print_debug(); // Output: Message1

	m1 = m1;
	m1.print_debug(); // Output: Message1
	f1.print_debug(); // Output nothing!

	return 0;
}

You'll see what I mean.

**Comment by pezy** [comment]
@clarelsy  I see. Thx a lot.

**Comment by clarelsy** [comment]
Could you please merge #451 too if you think it is okay?
#451 is the header file for #452... I should have put #451 and #452 in the same patch but not sure how to do that.

**Comment by pezy** [comment]
@clarelsy np.

**Pull request merged**